workpackage creation frontend

0xF013 opened the pull request:
https://community.openproject.org/work_packages/17549
- [ ] fix autocomplete noscope
- [x] refactor or remove task dropdown menu controller
- [x] fix refresh for /create url
- [x] fix annoying focus on each element on create mode
- [x] focus to title on create toggle
- [x] new work package header in german
    button_save_all: "Save"
    button_save_and_send_all: "Save and send email"
    button_cancel_all: "Cancel"
- [x] column arrows position bug
- [x] time group is now always empty
- [x] payload API title or name
- [x] type switch during creation

NobodysNightmare committed:
specify endpoint for current watchers

[ci skip]

NobodysNightmare committed:
require permissions to see watchers

[ci skip]

NobodysNightmare committed:
clarify 403/404 for WP scoped resources

[ci skip]

NobodysNightmare committed:
change the way watchers are added

- using links
- possibility for multiple watchers at a time
- always returns all watchers (pagination stilll possible)

Note that the term "link object" is introduced by an as of now pending
PR

[ci skip]

NobodysNightmare committed:
specify further error cases

[ci skip]

NobodysNightmare committed:
fix example link

[ci skip]

NobodysNightmare committed:
bring available watchers in line with assignees etc

- endpoint also scoped by project
- thus also returning all possible watchers, not only the ones "not
watching"

[ci skip]

NobodysNightmare committed:
clarification on 403/404 of available xxx

[ci skip]

NobodysNightmare committed:
also return 422 for non-watchable users

[ci skip]

NobodysNightmare committed:
Merge remote-tracking branch 'origin/dev' into fix/specify_watchers_link

Conflicts:
	doc/apiv3-documentation.apib

[ci skip]

NobodysNightmare committed:
rename data to payload, reference it in "links"

[ci skip]

NobodysNightmare committed:
only allow to add one watcher at a time

NobodysNightmare committed:
specify error id for type mismatch of links

[ci skip]

NobodysNightmare committed:
fixed example

[ci skip]

myabc committed:
Remove livingstyleguide as Rails app dependency

NOTE: removes the ability to build the style guide with Sprockets.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Upgrade to livingstyleguide v2 syntax

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Rename style guide doc files from .md → .lsg

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Add separate frontend Gemfile w/ livingstyleguide

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Bump gulp-livingstyleguide to 0.1.5

This version now uses `bundle exec` when calling out to the
`livingstyleguide` binary. See:
https://github.com/efacilitation/gulp-livingstyleguide/commit/d645828f591d453d15fac465b63956aa72882348

myabc committed:
[WORKAROUND] Add coffee-script as devDependency

Work around an issue with gulp-livingstyleguide packaging:
See https://github.com/efacilitation/gulp-livingstyleguide/issues/4

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Add html[id] to style guide based on file path

Required in order for the style guide's own styles to be correctly
scoped. File path is most likely temporary however. See:
https://github.com/livingstyleguide/livingstyleguide/issues/133

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Tell gulp to watch .lsg files instead

Signed-off-by: Alex Coles <alex@alexbcoles.com>

Commit:
Unified permission checks

Commit:
Bring activity_representer in line with the others

Commit:
Draft new work package create endpoints

Commit:
Fix activity user

Commit:
Swap work package creation links

Commit:
Extract shareable methods into separate module

Commit:
Increase service flexibility

Commit:
Introduce create work package endpoint

Commit:
Quieten the hound

Commit:
Collect writable attributes from ancestors

Commit:
Moved lock_version validation enforcement

Commit:
Provide more flexible structure for work package contract validations

Commit:
Appease the hound

Commit:
Introduce CreateWorkPackageService

Commit:
Improve initialization readability

Commit:
Move and rename AvailableWorkPackages API to meet standards

Commit:
Prepare WorkPackagesByProject for form endpoint

Commit:
Merge form folder into work_packages folder

Commit:
Improve refactorability

Commit:
Extract *_form_api endpoints

Commit:
Calm the hound

Commit:
Bump version to 4.3.0-alpha

Commit:
Accept projects as render context

NobodysNightmare committed:
Merge branch 'release/4.1' into dev

Commit:
Remove unnecessary destruction

NobodysNightmare committed:
Merge branch 'release/4.2' into dev

Commit:
Separating form representer for update and create cases

NobodysNightmare committed:
Merge branch 'release/4.2' into dev

Commit:
Consider contract information across the entire inheritance hierarchy

kgalli committed:
Merge branch 'release/4.2' into dev

Commit:
Adapt form representer tests

Commit:
Fix broken activity self_link

Commit:
Moved wp payload representer accordingly

Commit:
Some style fixes

Commit:
Fix permissions + cleanup

Commit:
Simplify work package services

Commit:
Increase type safety ;)

Commit:
Make writable_attributes public

Commit:
Correct work_package_contract filename

NobodysNightmare committed:
Merge branch 'release/4.2' into dev

NobodysNightmare committed:
Merge branch 'release/4.2' into fix/specify_watchers_link

Conflicts:
	doc/apiv3-documentation.apib

NobodysNightmare committed:
return ResourceTypeMismatch for wrong resources

- this is according to new specification changes
- I have no clue why the initial author did this "wrap into a validation error" business

NobodysNightmare committed:
adapt watch/unwatch links to specification

- they are called (un)watch, not (un)watchChanges
- watch link indicates the required payload using the payload attribute
- required payload already indicated using new (as of this commit invalid) format

NobodysNightmare committed:
remove redundant authentication

already performed on project level

NobodysNightmare committed:
style resource link parser

0xF013 committed:
code extraction

NobodysNightmare committed:
update roar version

NobodysNightmare committed:
extract parsable link into own representer

NobodysNightmare committed:
accept new request format for add watcher

also adapt the angular frontend to speak the new format

NobodysNightmare committed:
avoid useless JSON roundtrip

know your library they said...

NobodysNightmare committed:
fix specs to respect new error identifier

0xF013 committed:
add missing Y

Commit:
Fix typo

NobodysNightmare committed:
clarify more edge cases regarding watchers

NobodysNightmare committed:
try to improve readability of watcher spec

incl. Rubocop fixes

Commit:
Remove erroneous existent project_id property

NobodysNightmare committed:
allow to view list of current watchers

Commit:
Fix spec once again

NobodysNightmare committed:
fix HTTP 500 in case of not allowed watcher

NobodysNightmare committed:
also use named parameters for RemoveWatcher service

NobodysNightmare committed:
specify and test edge cases for watcher deletion

reconsidered when to return 404:

while the watcher resource does technically not exist when a user is not watching,
it seems more client friendly to still return a success code even if nothing was done

Still I differentiated that from user that does not exist at all

NobodysNightmare committed:
further specification refinements

- remove build warnings
- move available watchers endpoint back to WP scope

NobodysNightmare committed:
correctly link & embed watchers in WP

floriank committed:
Merge pull request #3048 from 0xF013/feature/code-stabilization-for-datepickers

datepickers code stabilization

Commit:
Merge branch 'release/4.2' into dev

Commit:
Merge branch 'release/4.2' into dev

Commit:
Choose only valid types

Commit:
µstyle fix

Commit:
Improve shared helper encapsulation

Commit:
Tiny style fix

NobodysNightmare committed:
Merge branch 'release/4.2' into dev

Commit:
Add all the specs

Commit:
Remove unused property

Commit:
Clarify comment

Commit:
Add condition to else

Commit:
Unify activity path declaration

Commit:
Use implicit getter

Commit:
Shorten form_representer by inheriting from single

Commit:
Add complimentary post

Commit:
Remove duplicated permission check

Commit:
Move service creation into post

Commit:
Remove superfluous variable

Commit:
Style fixes

Commit:
Correct wp create endpoint refactoring

Commit:
Adjust spec to service change

Commit:
Lull the hound

Commit:
Remove activity self link title

Commit:
Use proper inherited variable

Commit:
Move helper method to helpers

Commit:
Add specs for create form api

Commit:
Explicitly specify PayloadRepresenter

Commit:
Fix dumb mistake

Commit:
Remove outdated test

NobodysNightmare committed:
add spec for new link object

NobodysNightmare committed:
Merge branch 'release/4.2' into dev

ulferts committed:
Merge branch 'release/4.2' into dev

Commit:
Merge branch 'release/4.2' into dev

Commit:
Properly use caching mechanism

Commit:
Remove debug remnant

Commit:
Diversify tests

Commit:
Remove model_ prefix

Commit:
Cleanup test

Commit:
Clarify spec purpose

Commit:
Improve test setup and checks

Commit:
Ditch regex for more readable string

Commit:
Check link content as well

Commit:
Move test to proper location and benefit from it

ulferts committed:
Merge branch 'release/4.2' into dev

Commit:
Refactor test to use helper_class

Commit:
Simplify test setup

Commit:
Reduce code duplication

Commit:
Lower expectations

Commit:
Add tests for notification case

NobodysNightmare committed:
Merge remote-tracking branch 'upstream/release/4.2' into fix/specify_watchers_link

Conflicts:
	config/locales/de.yml
	config/locales/en.yml
	spec/lib/api/v3/work_packages/work_package_representer_spec.rb

Commit:
Diversify tests

kgalli committed:
Merge branch 'release/4.2' into dev

Commit:
Move out of place parenthesis back into position

NobodysNightmare committed:
simplify conditions + specs

if a user is not allowed to view a work package he should not get to see
the WP representer. (even hypothetically: then watcher links would not be the prime problem)

Therefore the conditions can be simplified as well as the corresponding specs

NobodysNightmare committed:
remove hardcoded titles

titles are specified as optional, but when they are present they should be localized

NobodysNightmare committed:
also provide a payload for addWatcher

allow payloads to be templated (to some degree) as well

Commit:
pin select 3.5.2

NobodysNightmare committed:
fix toggle watch code

Commit:
Add test descriptions

Commit:
Correct test description

Commit:
Remove redundant context

Commit:
Remove temporal coupling

0xF013 committed:
barely breathing wp form

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://0xf013-create-work-package-frontend.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://0xf013-create-work-package-frontend.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

0xF013 commented on frontend/app/work_packages/directives/inplace_editor/inplace-editor-edit-pane-directive.js:
```diff
@@ -182,13 +185,15 @@ module.exports = function(
-      });
+      if (!EditableFieldsState.forcedEditState) {
+        element.bind('keydown keypress', function(e) {
+          if (e.keyCode == 27) {
```
@houndci and I expected to be an officer on USS Enterprise when I grow up

ulferts committed:
Merge pull request #2990 from Azure7111/feature/create_work_package

Add create work package endpoints

Commit:
Allow raising exceptions from notification jobs

machisuji committed:
Merge pull request #3126 from barhot/feature/allow-failing-notification-jobs

Allow raising exceptions from notification jobs

ulferts committed:
Merge branch 'release/4.2' into dev

0xF013 committed:
buttons and functional view

Commit:
Merge branch 'release/4.2' into dev

ulferts committed:
Merge branch 'release/4.2' into dev

Commit:
Merge branch 'release/4.2' into dev

0xF013 committed:
seamless edit/create

myabc committed:
Bump rspec, rspec-rails to latest ~> 3.3.0

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix running of specs with rspec executable

Legacy specs should be filtered out by default when using `rspec`
executable with no arguments or a path that contains both legacy and
non-legacy specs (e.g. `rspec spec`).

As noted in 1a0e2b90, suites must be run separately for time-being.

[ci skip]

Signed-off-by: Alex Coles <alex@alexbcoles.com>

ulferts committed:
Merge pull request #3143 from myabc/fix/specs-rspec-executable

Fix running of specs with rspec executable

0xF013 committed:
cancel button buff

Commit:
Merge branch 'release/4.2' into dev

ulferts committed:
Merge pull request #3142 from myabc/feature/rspec-3.3

RSpec 3.3 Upgrade

0xF013 committed:
peace between create and update

ulferts committed:
Merge branch 'release/4.2' into dev

ulferts committed:
Merge branch 'release/4.2' into dev

0xF013 committed:
fix small visual issues

NobodysNightmare committed:
rename query to scope

sounds more understandable to me, but I am biased, I guess...

Commit:
Merge branch 'release/4.2' into dev

Commit:
Merge branch 'release/4.2' into dev

NobodysNightmare committed:
use more clear scopes for autocompletion

lets see if that really makes stuff more understandable...
I (personally) found the old definition of "all" to be too narrow.

Now there is:

scope == "all": I want to search over all WPs
scope == "relatable": I want to search over all WPs that are relatable to WPs in the given project (via project_id)
no scope: If I specified a project search over the WPs of that one, otherwise search everywhere (like "all")

NobodysNightmare committed:
explain why some code exists

floriank commented:
I found some small issues so far:
- somehow the assignee get's two empty options for me.
- the details pane spinner is gone when doing form submission, etc.
- switching from Edit to Create causes the pane to disappear temporarily

Nothing that should prevent an initial merge and a presentation I guess.

NobodysNightmare committed:
try to increase readability

- extract scopedeterminition into an extra method
- always use "outer" if statements in index method

NobodysNightmare committed:
always return a valid scope

though WorkPackage works like a scope *most* of the time, it is not a scope

e.g. you can't directly call :order on WorkPackage

Note: I would have loved to already do it the Rails 4 way and call :all
However in Rails 3 :all returns an Array and not a Relation, so we will
have to wait for the Rails 4(.1) Upgrade to replace :scoped with :all :(

Sorry @myabc

NobodysNightmare committed:
rewrite find into where, order and limit

NobodysNightmare committed:
increase readability of wp->hash conversion

it might use more lines, but I hope there is a gain in readability

ulferts committed:
Merge branch 'release/4.2' into dev

NobodysNightmare committed:
Merge pull request #3120 from floriank/fix/pin-select-version

[18714] Pin select2 3.5.2

NobodysNightmare committed:
Merge remote-tracking branch 'upstream/release/4.2' into fix/specify_watchers_link

Conflicts:
	lib/api/v3/work_packages/form/work_package_attribute_links_representer.rb
	lib/api/v3/work_packages/work_package_representer.rb
	lib/api/v3/work_packages/work_packages_api.rb

NobodysNightmare committed:
Merge pull request #2957 from myabc/feature/livingstyleguide-upgrade

NobodysNightmare committed:
eliminate instance variable

NobodysNightmare committed:
also accept auto complete requests without query

NobodysNightmare committed:
specify order in the Rails 3 way

ascending is the default anyway

ulferts committed:
Merge branch 'release/4.2' into dev

NobodysNightmare committed:
uglify order statement

at least this one DOES work...

ulferts committed:
Merge branch 'release/4.2' into dev

NobodysNightmare committed:
use "relatable" scope for relations

NobodysNightmare committed:
centralize knowledge of WP autocomplete url

- project information is not needed anymore
- the endpoint format was fixed anyway (also accepted parameters)

ulferts committed:
Merge branch 'release/4.2' into dev

Commit:
Merge branch 'release/4.2' into dev

0xF013 committed:
fix translations

0xF013 commented:
@floriank 
- That's not an empty options, rather a styling issue that accompanies the empty option. Might be worth a high prio premium WP
- The spinner's role ATM is taken by each individual field's blocked state. Can be changed though
- I can't pinpoint the reason for the pane behaviour at this moment

Commit:
Merge branch 'release/4.2' into dev

Commit:
add styles for alert messages

floriank commented:
> somehow the assignee get's two empty options for me.

That one is my bad - I actually have a user with the name 

``` html
<script>alert(0);</script>
```

for first and last name.

Commit:
add a button in alert message

Commit:
add simple notification message

Commit:
Merge remote-tracking branch 'opf/release/4.2' into release/4.3

NobodysNightmare committed:
fix c&p error

[ci skip]

NobodysNightmare committed:
add a removeWatcher link to work packages

NobodysNightmare committed:
fix spec

the actual hashes provided by grape have string keys

NobodysNightmare committed:
clarify specification

NobodysNightmare committed:
style

[ci skip]

floriank commented:
@0xF013 after the #3162 has been merged, the branch should probably rebased and the autocompletion stuff can be done, right? right? :cold_sweat:

0xF013 commented:
@floriank in theory

Commit:
add default font color for message box

Commit:
remove some spaces

NobodysNightmare committed:
fix doc error caused by moving the endpoint

endpoint was moved from project scope to WP scope

[ci skip]

NobodysNightmare committed:
Merge remote-tracking branch 'upstream/release/4.2' into fix/specify_watchers_link

Conflicts:
	spec/requests/api/v3/watcher_resource_spec.rb

NobodysNightmare committed:
clarify words

[ci skip]

NobodysNightmare committed:
style

[ci skip]

NobodysNightmare committed:
typo

[ci skip]

Commit:
Merge pull request #3115 from opf/fix/specify_watchers_link

NobodysNightmare committed:
actually allow the status to be set while moving

apparently the status_id param was set to expect an array of values while being moved in this commit: d7f399ab8db059153d2792f9db94450150c7ffdb
Probably because the spec started to expect that when it was moved to be an RSpec test.
I am not sure if this was ever a valid thing outside the tests...

NobodysNightmare committed:
fix moving of WPs without specifying a status

attributes without a value are actually represented by empty strings.
Thus only take the status_id if it is :present? (the inverse of :blank?)

NobodysNightmare committed:
Merge branch 'release/4.2' into release/4.3

NobodysNightmare committed:
Merge branch 'release/4.2' into release/4.3

Conflicts:
	lib/api/v3/errors/error_representer.rb

NobodysNightmare committed:
include missing attributes in mails

Commit:
change default notiffication message style

Azure7111 committed:
Merge pull request #3189 from NobodysNightmare/feature/change_status_during_move

fix changing status during WP move/copy

NobodysNightmare committed:
fix whitespacing in '_issue_details.text.erb'

empty lines and indentation used for code formatting is immediately visible in the resulting output.

I tried to employ some special rules to at least allow for usage of indentation.

Commit:
remove unused variables

ulferts committed:
Merge pull request #3192 from NobodysNightmare/feature/show_dates_in_mail

include missing attributes in mails

ulferts committed:
update webpack to avoid karma fails

myabc committed:
Merge pull request #3198 from ulferts/fix/karma_run

update webpack to avoid karma fails

ulferts committed:
Merge branch 'release/4.2' into release/4.3

NobodysNightmare committed:
Merge pull request #3165 from vcostin/feature/20580-errors-and-notifications

NobodysNightmare committed:
Merge pull request #3193 from myabc/feature/rubocop-semantic-blocks

Conflicts:
	Gemfile.lock

Note that this adds BUNDLED_WITH to your Gemfile.lock, in case this line gets lost when executing bundle,
make sure your local version of bundler is updated first ;-)

floriank commented:
rebased by me against the current `release/4.3`

Commit:
fix routing for index action

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add missing form mocks for tests

Signed-off-by: Florian Kraft <f.kraft@finn.de>

floriank commented:
Closed in favour of #3204 (reopenend against the release branch)
